crates/core/src/backend/decrypt.rs

```diff
@@ -395,12 +411,41 @@ impl<C: CryptoKey> DecryptWriteBackend for DecryptBackend<C> {
             }
             None => self.key().encrypt_data(data)?,
         };
-        let id = hash(&data);
-        self.write_bytes(tpe, &id, false, data.into())
+        if self.extra_check {
+            let check_data = self.decrypt_file(&data_encrypted)?;
+            if data != check_data {
+                return Err(CryptBackendErrorKind::ExtraCheckFailed.into());
```

Comment:
Would rather call it `integrity check` or something?

```suggestion
                return Err(CryptBackendErrorKind::IntegrityCheckFailed.into());
```